### ISSUE TITLE
Caddy - cannot set up HTTP (without TLS) listener

### ISSUE THREAD

**Issue by KiNgMaR** [opnsense/plugins#4074]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [X] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [X] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [X] The title contains the plugin to which this issue belongs

**Describe the bug**
- Caddy has a concept of "addresses" which is explained here, and among other things, is used to define whether a listening port is HTTP or HTTPS: https://caddyserver.com/docs/caddyfile/concepts#addresses
- To set up an HTTP listener on 10.123.0.11:8083, the Caddyfile should contain `http://10.123.0.11:8083`
- In the OPNSense plugin for Caddy, the `FromDomain` field does not allow entering hosts with protocol prefix, nor does a separate option exist that would allow creating a Caddy `address` that listens via HTTP.

**Screenshots**
![image](https://github.com/opnsense/plugins/assets/438358/9ca65b3b-871c-4334-83a1-c51ca6629520)

**Additional context**
Caddyfile (as is)
```
# Reverse Proxy Domain: "ce6e1fa0-2ffe-41ff-8d38-19fb8d39036c"
10.123.0.11:8083 {
	log ce6e1fa0-2ffe-41ff-8d38-19fb8d39036c

	handle @71c5b1ba-e561-44dd-9db0-4bdf4435da0f {
		...
	}
}
```

Caddyfile (expected)
```
# Reverse Proxy Domain: "ce6e1fa0-2ffe-41ff-8d38-19fb8d39036c"
http://10.123.0.11:8083 {
	log ce6e1fa0-2ffe-41ff-8d38-19fb8d39036c

	handle @71c5b1ba-e561-44dd-9db0-4bdf4435da0f {
		...
	}
}
```

**Environment**
OPNsense 24.1.9_4-amd64
FreeBSD 13.2-RELEASE-p11
OpenSSL 3.0.14
Caddy Auto HTTPS = OFF

**Comment by Monviech** [comment]
Hey there, I know what you mean.

It was a design choice of the plugin, to make sure there is no user error when setting up TLS for domains.

For use cases like ACME redirection, there is a predefined "http" section, that also keeps the redirect from http to https intact: https://github.com/opnsense/plugins/blob/master/www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile#L231

Right now, this field can't be changed anymore easily. If you explain the use case for not wanting TLS, I could provide a checkbox that says "Disable TLS" that is hidden under the advanced options of a domain, which will render "http://" in front of a domain. Not rendering anything in front of a domain uses "https://" as default without being explicitly rendered.

Right now, I can't imagine a use case where deactivating TLS makes sense, (Except the ACME Redirection I have shown).

**Comment by KiNgMaR** [comment]
Hey, thanks for the context, that was helpful. For me, the "Disable TLS" checkbox would do the trick.

My use case is as follows:
- the OPNSense router is connected to a bunch of VPNs, one of them being used to expose several endpoints from one location on the same host and port.
- since the VPN is already providing encryption and authentication, for this specific scenario, there's not really a point in using HTTPS-secured connections "inside" the VPN.

I suppose I could use the DNS challenge to actually have Caddy issue certificates, but would like to avoid the additional complexity until it's really needed.

Thanks for maintaining this btw, it's much appreciated.

**Comment by Monviech** [comment]
Yeah this makes sense, double encryption would be indeed useless in that scenario.

I'll make a PR ready next week.

That checkbox would also be logical.

- In the `reverse_proxy` directive, `http` is standard. And there is a checkbox named `tls` to activate `https`.

- In the `domain` block, `https` is standard, so a checkbox named `disable tls` would activate `http`.

And thanks, always nice to see people use it. :)

**Comment by Monviech** [comment]
A different thing. I just saw that you are using IP addresses in the domain field. Was that just for testing or is that going to be productive.

I just wanted to disallow IP addresses in the domains in the next release, since only SNI with domain names makes sense in an environment like the OPNsense. Is that a mistake in your case here?

https://github.com/opnsense/plugins/blob/4f48d1e7885d9cadf7aa416b68ca371a14ceae37/www/caddy/src/opnsense/mvc/app/models/OPNsense/Caddy/Caddy.xml#L154

**Comment by KiNgMaR** [comment]
Hehe, I was assuming that using an IP would also make Caddy `bind()` to that specific interface, but actually it doesn't ... maybe another feature request? Expose https://caddyserver.com/docs/caddyfile/directives/bind ?

TBH I don't really see the point in disallowing IPs in the domain field, but it would not break my usecase, can easily set up a wildcard DNS instead.

**Comment by Monviech** [comment]
I actually don't really know what an IP would do in the domain field. Without SNI there could only be one service per port, kinda defeating what a reverse proxy is for. (If I understood it right).

I don't think it binds to the interface IP address directly either.

I won't allow binding to a certain IP/interface in the GUI since there are a lot of drawbacks. You can configure it though in the shell, I wrote a guide here:

https://docs.opnsense.org/manual/how-tos/caddy.html#advanced-bind-caddy-to-specific-interface

**Comment by KiNgMaR** [comment]
> The interface won't be there at boot time and Caddy would fail to start.

Good point 👍 

>Without SNI there could only be one service per port

In my setup, different services are offered on path prefixes, using `handle_path`.

**Comment by Monviech** [comment]
I won't reintroduce the IP address in domains since its use is very unclear.

I will add the "deactivate tls" checkbox though. So, with an internal domain name where Caddy listens you'll reach the same goal.

**Comment by KiNgMaR** [comment]
Installed the version, updated the settings, and am happy to report that it works exactly as intended. Can finally uninstall the nginx plugin 😅

**Comment by Monviech** [comment]
Nice, in your case also look forward to this, which also comes soon:

https://github.com/opnsense/plugins/pull/4081

It synergizes with your port settings.
